File: java/src/com/szklimowicz/organisms/animals/Animal.java
package com.szklimowicz.organisms.animals;

import com.szklimowicz.Field;
import com.szklimowicz.World;
import com.szklimowicz.organisms.Organism;

import java.util.concurrent.ThreadLocalRandom;

public abstract class Animal extends Organism {

    public Animal(int x, int y, World world) {
        super(x, y, world);
        setName("Animal");
        setStrength(1);
        setInitiative(1);
    }

    @Override
    public void action() {

        setAge(getAge() + 1);
        int new_x = getX(), new_y = getY();
        int range = getRange();
        int dir = ThreadLocalRandom.current().nextInt(1, 4 + 1);

        switch(dir) {

            case LEFT:
                new_x -= range;
                break;
            case RIGHT:
                new_x += range;
                break;
            case UP:
                new_y -= range;
                break;
            case DOWN:
                new_y += range;
                break;
        }

        World world = getWorld();
        Class animal_class = this.getClass();

        if(world.inWorldBorders(new_x, new_y)) {
            Field target_field = world.getFields()[new_y][new_x];
            if (target_field.getOrganism() == null) {
                target_field.setOrganism(this);
                world.getFields()[getY()][getX()].setOrganism(null);
                setX(new_x);
                setY(new_y);
            }
            else if (animal_class.isInstance(target_field.getOrganism())) {
                breed(animal_class, world);
            }
            else collisionAttack(this, target_field.getOrganism());
        }
    }

    private void breed(Class animal_class, World world) {
        if(Math.random() > 0.5) {

            int new_x = getX(), new_y = getY();
            int dir = ThreadLocalRandom.current().nextInt(1, 4 + 1);

            switch(dir) {

                case LEFT:
                    new_x--;
                    break;
                case RIGHT:
                    new_x++;
                    break;
                case UP:
                    new_y--;
                    break;
                case DOWN:
                    new_y++;
                    break;
            }

            if(world.inWorldBorders(new_x, new_y)) {
                if(world.getFields()[new_y][new_x].getOrganism() == null) {
                    try {
                        Organism org = (Organism) animal_class.getDeclaredConstructor(int.class, int.class, World.class).newInstance(new_x, new_y, world);
                        world.addNewOrganism(org);
                    }
                    catch (Exception ex) {
                        ex.printStackTrace();
                    }
                }
            }
        }
    }
}

File: python/animal.py
from organism import Organism
from random import randint

class Animal(Organism):
        
    def __init__(self, x, y, world):
        super().__init__(x, y, world)
        self.name = "Animal"

    def action(self, null):

        dir = randint(1,4)

        new_x = self.x
        new_y = self.y

        if(dir == 1):
            new_y = self.y - self.moverange
        if(dir == 2):
            new_y = self.y + self.moverange
        if(dir == 3):
            new_x = self.x - self.moverange
        if(dir == 4):
            new_x = self.x + self.moverange


        if self.world.inWorld(new_x, new_y):
            if type(self) == type(self.world.organisms[new_y][new_x]):
                self.breed(type(self))
                return
            else:
               if self.world.organisms[new_y][new_x] == 0:
                   self.world.organisms[new_y][new_x] = self.world.organisms[self.y][self.x]
                   self.world.organisms[self.y][self.x] = 0
                   self.x = new_x
                   self.y = new_y
                   self.ID = new_x*23 + new_y
               else:
                   if self.collision(self, self.world.organisms[new_y][new_x]):
                      return
                   else:
                       return

    def breed(self, type):

        breedchance = randint(0,1)

        if breedchance:
            dir = randint(1,4)

            new_x = self.x
            new_y = self.y

            if(dir == 1):
                new_y = self.y - self.moverange
            if(dir == 2):
                new_y = self.y + self.moverange
            if(dir == 3):
                new_x = self.x - self.moverange
            if(dir == 4):
                new_x = self.x + self.moverange
        
            if self.world.inWorld(new_x, new_y):
                if self.world.organisms[new_y][new_x] == 0:
                    self.world.addNewOrganism(type, new_x, new_y)